caldav/match.go

```diff
@@ -0,0 +1,224 @@
+package caldav
+
+import (
+	"time"
+
+	"github.com/emersion/go-ical"
+)
+
+// Filter returns the filtered list of calendar objects matching the provided query.
+// A nil query will return the full list of calendar objects.
+func Filter(query *CalendarQuery, cos []CalendarObject) ([]CalendarObject, error) {
+	if query == nil {
+		// FIXME: should we always return a copy of the provided slice?
+		return cos, nil
+	}
+
+	out := make([]CalendarObject, 0)
+	for _, co := range cos {
+		ok, err := Match(query.CompFilter, &co)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			continue
+		}
+
+		// TODO properties are not currently filtered even if requested
+		out = append(out, co)
+	}
+	return out, nil
+}
+
+// Match reports whether the provided CalendarObject matches the query.
+func Match(query CompFilter, co *CalendarObject) (matched bool, err error) {
+	if co.Data == nil || co.Data.Component == nil {
+		// TODO is this a thing? Should this match if comp.IsNotDefined?
+		return false, nil
+	}
+	comp := co.Data.Component
+
+	if query.Name != comp.Name {
+		return query.IsNotDefined, nil
+	}
+
+	for _, compFilter := range query.Comps {
+		match, err := matchCompFilter(compFilter, comp)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+	}
+	for _, propFilter := range query.Props {
+		match, err := matchPropFilter(propFilter, comp)
+		if err != nil {
+			return false, err
+		}
+		if !match {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func matchCompFilter(filter CompFilter, comp *ical.Component) (bool, error) {
+	var matches []*ical.Component
+
+	for _, child := range comp.Children {
+		match, err := matchCompFilterChild(filter, child)
+		if err != nil {
+			return false, err
+		} else if match {
+			matches = append(matches, child)
+		}
+	}
+	if len(matches) == 0 {
+		return filter.IsNotDefined, nil
+	}
+	return true, nil
+}
+
+func matchCompFilterChild(filter CompFilter, comp *ical.Component) (bool, error) {
```

Comment:
Hm. It seems like this duplicates code with `Match`. Can't we remove the root component special case?

Reply:
Thanks for nudging me on this, the recursion is much cleaner now.